File: problems/Minimum Path Sum/minPathSum.cs
public class Solution {
    public int MinPathSum(int[][] grid) {
        var dp = new int[grid[0].Length];

        for (int i = grid.Length - 1; i >= 0; i--) {
            for (int j = grid[0].Length - 1; j >= 0; j--) {
                if(i == grid.Length - 1 && j != grid[0].Length - 1)
                    dp[j] = grid[i][j] +  dp[j + 1];
                else if(j == grid[0].Length - 1 && i != grid.Length - 1)
                    dp[j] = grid[i][j] + dp[j];
                else if(j != grid[0].Length - 1 && i != grid.Length - 1)
                    dp[j] = grid[i][j] + Math.Min(dp[j], dp[j + 1]);
                else
                    dp[j] = grid[i][j];
            }
        }

        return dp[0];
    }
}

File: problems/Minimum Path Sum/minPathSum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int[] dp = new int[cols];

        for (int row = 0; rows > row; ++row) {
            for (int col = 0; cols > col; ++col) {
                if (0 == row && 0 == col) {
                    dp[col] = grid[row][col];
                } else  {
                    int fromUp = 0 < row ? dp[col] : Integer.MAX_VALUE;
                    int fromLeft = 0 < col ? dp[col - 1] : Integer.MAX_VALUE;

                    dp[col] = grid[row][col] + Math.min(fromUp, fromLeft);
                }
            }
        }

        return dp[cols - 1];
    }
}
